ENH: quadratic assignment problem

pmla opened the pull request:
#### What does this implement/fix?
Building off the great work of @asaadeldin11 in #11862, I have implemented a bare-bones method for approximate solution of the quadratic assignment problem. The method implements [Umeyama's eigendecomposition approach](https://doi.org/10.1109/34.6778). This method is simple, deterministic, and parameter-free.

#### Additional information
I have tried to get 100% code coverage with the tests. There are no benchmarks included, since they can be added at a later stage.

I have tried to copy the interface of `linear_sum_assignment` as much as possible. Calculating the score is perhaps a bit unwieldy though. I have left `row_ind` in the return value so we can implement rectangular matrices at some point.

mdhaber commented:
I'm afraid I can't review this fully right now, but since this was inspired by gh-11862, I think it should be worked into gh-11862 rather than forcing the authors of the original PR to rework their contribution into this. Does that make sense? 
Also, if we are going to add this as a separate method, then didn't you want there to be a `method` argument @pmla? Shouldn't `quadratic_assignment` just dispatch to another function depending on the value of `method`?
I worked with the authors of gh-11862 quite a bit on the interface, and while not everything was finished, one decision that was important was that the method should not simply return `row_ind, col_ind` like `linear_sum_assignment` because that cannot be extended in the future. It should return an `OptimizeResult`.

pmla commented:
Thanks for your comments Matt.

> since this was inspired by gh-11862, I think it should be worked into gh-11862 rather than forcing the authors of the original PR to rework their contribution into this. Does that make sense?

#11862 is very large, and working this into it will make it larger. My intention with this PR was to make a minimal change set which can be built upon with later PR's. If that is no good, we can close this and continue with #11862 as is.

> if we are going to add this as a separate method, then didn't you want there to be a method argument @pmla? Shouldn't quadratic_assignment just dispatch to another function depending on the value of method?

I have split the solution logic off into a separate file, `_qap_umeyama.py`. As more solution methods are added, the logic of `_qap.py` can be extended to dispatch to them. A `method` argument is necessary until more methods are added.

> I worked with the authors of gh-11862 quite a bit on the interface, and while not everything was finished, one decision that was important was that the method should not simply return row_ind, col_ind like linear_sum_assignment because that cannot be extended in the future. It should return an OptimizeResult.

I have changed the return value to `OptimizeResult`. I also copied the examples from the docstring in #11862.

ilayn commented on scipy/optimize/_qap_umeyama.py:
```diff
@@ -0,0 +1,17 @@
+def eigenvectors(x):
+    # get eigenvectors sorted by eigenvalue magnitude (large to small)
+    l, v = np.linalg.eig(x)
+    indices = np.argsort(l, kind='merge')[::-1]
```
is X always hermitian? otherwise you can't sort reliably. If it is, then use `eigh` from scipy which auto-sorts them.

Also why `merge`?

pmla replied:
Thanks for pointing that daft mistake out. The matrices are not in general hermitian.

I have added handling and tests for non-hermitian matrices. The `eigenvectors` function has been replaced with plain old `eigh`.

ilayn commented on scipy/optimize/_qap_umeyama.py:
```diff
@@ -0,0 +1,17 @@
+from . import linear_sum_assignment, OptimizeResult
+
+
+def eigenvectors(x):
```
This function looks too small to refactor I think. In fact this file is tiny to get separated for my taste but it's not a blocker.

pmla replied:
I'll put it back in `_qap.py`.

ilayn commented on scipy/optimize/_qap_umeyama.py:
```diff
@@ -0,0 +1,17 @@
+    # Finds an approximate solution to QAP using Umeyama's method
+    W = np.abs(eigenvectors(B)) @ np.abs(eigenvectors(A)).T
+    _, col_ind = linear_sum_assignment(W, maximize=maximize)
+    score = float(np.trace(A @ B[col_ind][:, col_ind].T))
```
I don't know if it is required for performance but `np.sum(A*B[ind][:, ind])` would be ~4x faster.

pmla replied:
I have used your suggested code and added a comment for readability.

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -0,0 +1,116 @@
+    A = np.asarray(A)
+    B = np.asarray(B)
+
+    if A.ndim != 2:
```
This doesn't check for squareness, right? It could be rectangular.

In any case, I'm going to merge this with gh-11862, so we'll take care of input validation there.

pmla replied:
Whoops. You are right.
I have added a check for squareness all the same so I can mark this as resolved.

ilayn commented on scipy/optimize/_qap.py:
```diff
@@ -0,0 +1,116 @@
+def umeyama_eigendecomposition(A, B, maximize=False):
+    """Finds an approximate solution to QAP using Umeyama's method."""
+    # 'Correct' non-hermitian matrices
+    if (A != A.T).any() or (B != B.T).any():
```
this won't catch complex arrays in case it is possible to push complex A, B

pmla replied:
It is not supposed to handle complex A, B.
I have added a check for this now.

ilayn commented on scipy/optimize/_qap.py:
```diff
@@ -0,0 +1,116 @@
+
+    va = np.linalg.eigh(A)[1]
+    vb = np.linalg.eigh(B)[1]
+    W = np.abs(vb) @ np.abs(va).T
```
Do you mean va.real and vb.real here? Why abs?

mdhaber replied:
Umeyama's paper takes the magnitude:
![image](https://user-images.githubusercontent.com/6570539/89695311-6226f300-d8c8-11ea-9b67-209da6bf3bc2.png)

pmla replied:
That is my understanding of it too.
Going to mark as resolved.

mdhaber commented:
I'm also noticing some unusual behavior... the score doesn't match the output matrix, and it's lower than the global optimum.

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -0,0 +1,116 @@
+    _, col_ind = linear_sum_assignment(W, maximize=maximize)
+
+    # faster version of `float(np.trace(A @ B[col_ind][:, col_ind].T))`
+    score = float(np.real(np.sum(A * B[col_ind][:, col_ind])))
```
Should this be calculated based on the original matrices?

mdhaber replied:
I'm not sure that's the only problem. Using these matrices:
```
A = [
    [0, 90, 10, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [90, 0, 0, 23, 0, 0, 0, 0, 0, 0, 0, 0],
    [10, 0, 0, 0, 43, 0, 0, 0, 0, 0, 0, 0],
    [0, 23, 0, 0, 0, 88, 0, 0, 0, 0, 0, 0],
    [0, 0, 43, 0, 0, 0, 26, 0, 0, 0, 0, 0],
    [0, 0, 0, 88, 0, 0, 0, 16, 0, 0, 0, 0],
    [0, 0, 0, 0, 26, 0, 0, 0, 1, 0, 0, 0],
    [0, 0, 0, 0, 0, 16, 0, 0, 0, 96, 0, 0],
    [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 29, 0],
    [0, 0, 0, 0, 0, 0, 0, 96, 0, 0, 0, 37],
    [0, 0, 0, 0, 0, 0, 0, 0, 29, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 37, 0, 0],
]
B = [
    [0, 36, 54, 26, 59, 72, 9, 34, 79, 17, 46, 95],
    [36, 0, 73, 35, 90, 58, 30, 78, 35, 44, 79, 36],
    [54, 73, 0, 21, 10, 97, 58, 66, 69, 61, 54, 63],
    [26, 35, 21, 0, 93, 12, 46, 40, 37, 48, 68, 85],
    [59, 90, 10, 93, 0, 64, 5, 29, 76, 16, 5, 76],
    [72, 58, 97, 12, 64, 0, 96, 55, 38, 54, 0, 34],
    [9, 30, 58, 46, 5, 96, 0, 83, 35, 11, 56, 37],
    [34, 78, 66, 40, 29, 55, 83, 0, 44, 12, 15, 80],
    [79, 35, 69, 37, 76, 38, 35, 44, 0, 64, 39, 33],
    [17, 44, 61, 48, 16, 54, 11, 12, 64, 0, 70, 86],
    [46, 79, 54, 68, 5, 0, 56, 15, 39, 70, 0, 18],
    [95, 36, 63, 85, 76, 34, 37, 80, 33, 86, 18, 0],
]
```
minimization finds a permutation with a higher score than maximization.

pmla replied:
> This should be calculated based on the original matrices, not the modified matrices. This affects the output when A and B are asymmetric.

Thanks. That should be fixed now.

mdhaber commented on scipy/optimize/tests/test_quadratic_assignment.py:
```diff
@@ -0,0 +1,85 @@
+
+@pytest.mark.parametrize("n", [10, 100])
+@pytest.mark.parametrize("directed", [False, True])
+def test_weighted_graph_isomorphism(n, directed):
```
I'll combine these tests with those from gh-11862, so no need to check that everything is covered.

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -0,0 +1,116 @@
+    >>> res = quadratic_assignment(A, B)
+    >>> print(res)
+     col_ind: array([3, 2, 1, 0])
+       score: 200.0
```
Of the 24 possible permutations, 6 have lower scores than this. Is it supposed to work better than this, or is that just how the algorithm does?

pmla replied:
This was a poor assumption on my part. It does not appear to work for minimization.
I will see if it can be adapted.

pmla replied:
Whilst the eigenvalue bounds work in both directions, Umeyama makes an assumption about isomorphism between the two graphs which means that the algorithm does not work for minimization.
I could raise a ValueError if `maximize=False` is passed. What do you think?

asaadeldin11 replied:
or we could send them to FAQ with a user warning

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -0,0 +1,116 @@
+        A = (A + A.T) / 2 + 1j * (A - A.T) / 2
+        B = (B + B.T) / 2 + 1j * (B - B.T) / 2
+
+    va = np.linalg.eigh(A)[1]
```
SciPy returns eigenvalues (and corresponding eigenvectors) in increasing order, opposite the convention in the paper. This seems to affect the output for asymmetric matrices.

pmla replied:
Nice suggestion. I have swapped the eigenvector ordering and renamed the variables.

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -0,0 +1,116 @@
+        B = (B + B.T) / 2 + 1j * (B - B.T) / 2
+
+    va = np.linalg.eigh(A)[1]
+    vb = np.linalg.eigh(B)[1]
```
```suggestion
    vb = np.linalg.eigh(B)[1][::-1]
```

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -0,0 +1,116 @@
+    va = np.linalg.eigh(A)[1]
+    vb = np.linalg.eigh(B)[1]
+    W = np.abs(vb) @ np.abs(va).T
+    _, col_ind = linear_sum_assignment(W, maximize=maximize)
```
After making the corrections above, `W` is correct. But for [2] (original paper) Section IV B, `linear_sum_assignment` is not giving the output reported in the paper.

pmla replied:
OK. I'll change this and add comments in the code/tests to explain the departure from the convention in the paper.

pmla replied:
I have changed this now.

mdhaber commented on scipy/optimize/tests/test_quadratic_assignment.py:
```diff
@@ -0,0 +1,85 @@
+                  [0, 1, 2, 0]])
+
+    result = quadratic_assignment(A, B, maximize=True)
+    assert_array_equal(result.col_ind, [0, 2, 3, 1])
```
```suggestion
    assert_array_equal(result.col_ind, [0, 3, 1, 2])
```
This is not the answer printed in Umeyama.
![image](https://user-images.githubusercontent.com/6570539/89611066-08271e80-d831-11ea-83c1-db45116986f8.png)

mdhaber replied:
@asaadeldin11 @jovo Do you have easy acess to an implementation of Umeyama's method? Can you check the results on this problem?

mdhaber replied:
I think Umeyama would recommend the transpose of the permutation we were using.

pmla replied:
> This is not the answer printed in Umeyama.

If we do:
```
>>> result = linear_sum_assignment(W, maximize=maximize)
>>> P = np.zeros(W.shape, dtype=int)
>>> P[result] = 1
>>> P
array([[1 0 0 0]
       [0 0 1 0]
       [0 0 0 1]
       [0 1 0 0]])
```

Since we have `row_ind, col_ind = result` I think this is consistent with Umeyama? Note that I am in agreement that Umeyama's permutation is transposed with respect to that of FAQ.

mdhaber replied:
Sorry, I edited my post. It was old.
It _is_ in agreement with Umeyama, but Umeyama solves the transposed problem (and for that, Umeyema notes that it is the global optimum). You'll find that it does not do as well for our problem as the transpose does (and you'll find that for our problem, this transpose is the global optimum).

pmla replied:
OK - we're on the same page.
I have changed the permutation convention. Also the superficial one you mention.

jovo commented:
i think iGraphMatch has an R implementation we could compare to:
https://rdrr.io/github/dpmcsuss/iGraphMatch/man/graph_match_methods.html


On Fri, Aug 7, 2020 at 1:47 AM Matt Haberland <notifications@github.com>
wrote:

>
> *      External Email - Use Caution      *
>
>
>
> *@mdhaber* commented on this pull request.
> ------------------------------
>
> In scipy/optimize/tests/test_quadratic_assignment.py
> <https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fscipy%2Fscipy%2Fpull%2F12161%23discussion_r466833742&data=02%7C01%7Cjovo%40jhu.edu%7C2608ccf699284540a0d408d83a955d8c%7C9fa4f438b1e6473b803f86f8aedf0dec%7C0%7C0%7C637323760479099123&sdata=qmzBslqejomSqGgiAUzt9WRaUtvzDWTKCh%2BcbRKBwTE%3D&reserved=0>
> :
>
> > +
> +def test_directed_weighted_graph_matching():
> +    # Example in section IVB of Umeyama:
> +    # "An eigendecomposition approach to weighted graph matching"
> +    A = np.array([[0, 3, 4, 2],
> +                  [0, 0, 1, 2],
> +                  [1, 0, 0, 1],
> +                  [0, 0, 1, 0]])
> +
> +    B = np.array([[0, 4, 2, 4],
> +                  [0, 0, 1, 0],
> +                  [0, 2, 0, 2],
> +                  [0, 1, 2, 0]])
> +
> +    result = quadratic_assignment(A, B, maximize=True)
> +    assert_array_equal(result.col_ind, [0, 2, 3, 1])
>
> @asaadeldin11
> <https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fasaadeldin11&data=02%7C01%7Cjovo%40jhu.edu%7C2608ccf699284540a0d408d83a955d8c%7C9fa4f438b1e6473b803f86f8aedf0dec%7C0%7C0%7C637323760479109128&sdata=IRiXi7jg3WCg7yPrq5Cdxuh%2BdvDUXUs7FuFwJqR6mBk%3D&reserved=0>
> @jovo
> <https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjovo&data=02%7C01%7Cjovo%40jhu.edu%7C2608ccf699284540a0d408d83a955d8c%7C9fa4f438b1e6473b803f86f8aedf0dec%7C0%7C0%7C637323760479109128&sdata=F8u2EYQIhN1GO5naAUd15FRmcuGI84DcIAXA2ydaxg4%3D&reserved=0>
> Do you have easy access to an implementation of Umeyama's method? Can you
> check the results on this problem?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fscipy%2Fscipy%2Fpull%2F12161%23discussion_r466833742&data=02%7C01%7Cjovo%40jhu.edu%7C2608ccf699284540a0d408d83a955d8c%7C9fa4f438b1e6473b803f86f8aedf0dec%7C0%7C0%7C637323760479119115&sdata=NlBXC1QlbFhXuCRzuDCWIC0AiEY3bXuSw1vaZZjX13c%3D&reserved=0>,
> or unsubscribe
> <https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAAAKG4QH6BKVIC2N6GLAKFLR7OIO3ANCNFSM4NEOM5BA&data=02%7C01%7Cjovo%40jhu.edu%7C2608ccf699284540a0d408d83a955d8c%7C9fa4f438b1e6473b803f86f8aedf0dec%7C0%7C0%7C637323760479119115&sdata=1nq9PrPEOBHnkPaWrEMTVIj2E1EAQawTZSWsaTBSv7c%3D&reserved=0>
> .
>


-- 
With gratitude,

Joshua T Vogelstein, PhD
neurodata.io | BME@JHU <https://www.bme.jhu.edu/> | dA/dt > 0 > dJ/dt
<https://twitter.com/neuro_data/status/1279067902658916352>
where A = appreciating, J = judging, and t = time

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -0,0 +1,116 @@
+    vb = np.linalg.eigh(B)[1]
+    W = np.abs(vb) @ np.abs(va).T
+    _, col_ind = linear_sum_assignment(W, maximize=maximize)
+
```
```suggestion
   col_ind = np.argsort(col_ind)
```
Umeyama's Hungarian algorithm looks for solutions to:
![image](https://user-images.githubusercontent.com/6570539/89694856-da8cb480-d8c6-11ea-8f8a-e6fc71965734.png)

Wherease `linear_sum_assignment`:
![image](https://user-images.githubusercontent.com/6570539/89694875-eb3d2a80-d8c6-11ea-8a9c-794f8c115d68.png)

Consequently our permutation matrix is the transpose of Umeyama's.

mdhaber commented on scipy/optimize/_qap.py:
```diff
@@ -2,20 +2,25 @@
+    else:
+        UG = np.linalg.eigh(AG)[1][:, ::-1]
+        UH = np.linalg.eigh(AH)[1][:, ::-1]
+        W = np.abs(UH) @ np.abs(UG).T
```
This looks great.

jovo commented on scipy/optimize/__init__.py:
```diff
@@ -255,12 +255,14 @@
    :toctree: generated/
 
    linear_sum_assignment -- Solves the linear-sum assignment problem.
+   quadratic_assignment -- Solves the quadratic assignment problem.
```
I'm not sure what the convention is here, but I think I would write

"Approximately solves the quadratic assignment problem"

And only say solves if we have an algorithm that finds the global optimal exactly with certainty?
